tools/pipeline.c

```diff
@@ -8,21 +8,53 @@
 
 #include "pipeline.h"
 
+#include <glib-unix.h>
 #include <gio/gio.h>
 #include <gst/gst.h>
 
+static guint signal_watch_intr_id;
+static guint target_id;
+
 static void
 activate (GApplication * app, gpointer user_data)
 {
   g_autoptr (GError) error = NULL;
   const gchar *fifo = NULL;
 
-
   GaeguliPipeline *pipeline = user_data;
 
-  fifo = g_object_get_data (G_OBJECT (app), "fifo");
-  gaeguli_pipeline_add_fifo_target (pipeline, fifo, &error);
   g_application_hold (app);
+
+  fifo = g_object_get_data (G_OBJECT (app), "fifo");
+  target_id = gaeguli_pipeline_add_fifo_target (pipeline, fifo, &error);
+}
+
+static void
+shutdown (GApplication * app, gpointer user_data)
+{
+  g_debug ("app released");
+  g_application_release (app);
```

Comment:
By the time `GApplication` fires its `"shutdown"` signal, it has been already released. I think you want to move `g_application_release (app);` to `stream_stopped_cb()`.

With that, this callback is unnecessary unless you want to keep the debug message.